### PR TITLE
[SPARK-28895][K8S] Support defining HADOOP_CONF_DIR and config map at the same time

### PR DESCRIPTION
**Pull request by yaooqinn**
### What changes were proposed in this pull request?

Changes in this pull request will support users to define `HADOOP_CONF_DIR` and `spark.kubernetes.hadoop.configMapName` at the same time. When both of them are defined, Spark will take precedence over the config map to be mounted on the driver pod. This enables the spark client process to communicate any Hadoop cluster if it needs.

### Why are the changes needed?

The *BasicDriverFeatureStep* for Spark on Kubernetes will upload the files/jars specified by --files/–jars to a Hadoop compatible file system configured by spark.kubernetes.file.upload.path. While using HADOOP_CONF_DIR, the spark-submit process can recognize the file system, but when using spark.kubernetes.hadoop.configMapName which only will be mount on the Pods not applied back to our client process. 
 
```scala
 Kent@KentsMacBookPro  ~/Documents/spark-on-k8s/spark-3.0.0-SNAPSHOT-bin-2.7.3  bin/spark-submit --conf spark.kubernetes.file.upload.path=hdfs://hz-cluster10/user/kyuubi/udf --jars /Users/Kent/Documents/spark-on-k8s/spark-3.0.0-SNAPSHOT-bin-2.7.3/hadoop-lzo-0.4.20-SNAPSHOT.jar --conf spark.kerberos.keytab=/Users/Kent/Downloads/kyuubi.keytab --conf spark.kerberos.principal=kyuubi/dev@HADOOP.HZ.NETEASE.COM --conf  spark.kubernetes.kerberos.krb5.path=/etc/krb5.conf  --name hehe --deploy-mode cluster --class org.apache.spark.examples.HdfsTest   local:///opt/spark/examples/jars/spark-examples_2.12-3.0.0-SNAPSHOT.jar hdfs://hz-cluster10/user/kyuubi/hive_db/kyuubi.db/hive_tbl
Listening for transport dt_socket at address: 50014
# spark.master=k8s://https://10.120.238.100:7443
19/08/27 17:21:06 WARN NativeCodeLoader: Unable to load native-hadoop library for your platform... using builtin-java classes where applicable
Using Spark's default log4j profile: org/apache/spark/log4j-defaults.properties
19/08/27 17:21:07 INFO SparkKubernetesClientFactory: Auto-configuring K8S client using current context from users K8S config file
Listening for transport dt_socket at address: 50014
Exception in thread "main" org.apache.spark.SparkException: Uploading file /Users/Kent/Documents/spark-on-k8s/spark-3.0.0-SNAPSHOT-bin-2.7.3/hadoop-lzo-0.4.20-SNAPSHOT.jar failed...
	at org.apache.spark.deploy.k8s.KubernetesUtils$.uploadFileUri(KubernetesUtils.scala:287)
	at org.apache.spark.deploy.k8s.KubernetesUtils$.$anonfun$uploadAndTransformFileUris$1(KubernetesUtils.scala:246)
	at scala.collection.TraversableLike.$anonfun$map$1(TraversableLike.scala:237)
	at scala.collection.mutable.ResizableArray.foreach(ResizableArray.scala:62)
	at scala.collection.mutable.ResizableArray.foreach$(ResizableArray.scala:55)
	at scala.collection.mutable.ArrayBuffer.foreach(ArrayBuffer.scala:49)
	at scala.collection.TraversableLike.map(TraversableLike.scala:237)
	at scala.collection.TraversableLike.map$(TraversableLike.scala:230)
	at scala.collection.AbstractTraversable.map(Traversable.scala:108)
	at org.apache.spark.deploy.k8s.KubernetesUtils$.uploadAndTransformFileUris(KubernetesUtils.scala:245)
	at org.apache.spark.deploy.k8s.features.BasicDriverFeatureStep.$anonfun$getAdditionalPodSystemProperties$1(BasicDriverFeatur# spark.master=k8s://https://10.120.238.100:7443
eStep.scala:165)
	at scala.collection.immutable.List.foreach(List.scala:392)
	at org.apache.spark.deploy.k8s.features.BasicDriverFeatureStep.getAdditionalPodSystemProperties(BasicDriverFeatureStep.scala:163)
	at org.apache.spark.deploy.k8s.submit.KubernetesDriverBuilder.$anonfun$buildFromFeatures$3(KubernetesDriverBuilder.scala:60)
	at scala.collection.LinearSeqOptimized.foldLeft(LinearSeqOptimized.scala:126)
	at scala.collection.LinearSeqOptimized.foldLeft$(LinearSeqOptimized.scala:122)
	at scala.collection.immutable.List.foldLeft(List.scala:89)
	at org.apache.spark.deploy.k8s.submit.KubernetesDriverBuilder.buildFromFeatures(KubernetesDriverBuilder.scala:58)
	at org.apache.spark.deploy.k8s.submit.Client.run(KubernetesClientApplication.scala:101)
	at org.apache.spark.deploy.k8s.submit.KubernetesClientApplication.$anonfun$run$10(KubernetesClientApplication.scala:236)
	at org.apache.spark.deploy.k8s.submit.KubernetesClientApplication.$anonfun$run$10$adapted(KubernetesClientApplication.scala:229)
	at org.apache.spark.util.Utils$.tryWithResource(Utils.scala:2567)
	at org.apache.spark.deploy.k8s.submit.KubernetesClientApplication.run(KubernetesClientApplication.scala:229)
	at org.apache.spark.deploy.k8s.submit.KubernetesClientApplication.start(KubernetesClientApplication.scala:198)
	at org.apache.spark.deploy.SparkSubmit.org$apache$spark$deploy$SparkSubmit$$runMain(SparkSubmit.scala:920)
	at org.apache.spark.deploy.SparkSubmit.doRunMain$1(SparkSubmit.scala:179)
	at org.apache.spark.deploy.SparkSubmit.submit(SparkSubmit.scala:202)
	at org.apache.spark.deploy.SparkSubmit.doSubmit(SparkSubmit.scala:89)
	at org.apache.spark.deploy.SparkSubmit$$anon$2.doSubmit(SparkSubmit.scala:999)
	at org.apache.spark.deploy.SparkSubmit$.main(SparkSubmit.scala:1008)
	at org.apache.spark.deploy.SparkSubmit.main(SparkSubmit.scala)
Caused by: java.lang.IllegalArgumentException: java.net.UnknownHostException: hz-cluster10
	at org.apache.hadoop.security.SecurityUtil.buildTokenService(SecurityUtil.java:378)
	at org.apache.hadoop.hdfs.NameNodeProxies.createNonHAProxy(NameNodeProxies.java:310)
	at org.apache.hadoop.hdfs.NameNodeProxies.createProxy(NameNodeProxies.java:176)
	at org.apache.hadoop.hdfs.DFSClient.<init>(DFSClient.java:678)
	at org.apache.hadoop.hdfs.DFSClient.<init>(DFSClient.java:619)
	at org.apache.hadoop.hdfs.DistributedFileSystem.initialize(DistributedFileSystem.java:149)
	at org.apache.hadoop.fs.FileSystem.createFileSystem(FileSystem.java:2669)
	at org.apache.hadoop.fs.FileSystem.access$200(FileSystem.java:94)
	at org.apache.hadoop.fs.FileSystem$Cache.getInternal(FileSystem.java:2703)
	at org.apache.hadoop.fs.FileSystem$Cache.get(FileSystem.java:2685)
	at org.apache.hadoop.fs.FileSystem.get(FileSystem.java:373)
	at org.apache.spark.util.Utils$.getHadoopFileSystem(Utils.scala:1881)
	at org.apache.spark.deploy.k8s.KubernetesUtils$.uploadFileUri(KubernetesUtils.scala:278)
	... 30 more
Caused by: java.net.UnknownHostException: hz-cluster10
	... 43 more
```

Other related spark configurations
```
spark.master=k8s://https://10.120.238.100:7443
# spark.master=k8s://https://10.120.238.253:7443
spark.kubernetes.container.image=harbor-inner.sparkonk8s.netease.com/tenant1-project1/spark:v3.0.0-20190813
# spark.kubernetes.driver.container.image=harbor-inner.sparkonk8s.netease.com/tenant1-project1/spark:v3.0.0-20190813
# spark.kubernetes.executor.container.image=harbor-inner.sparkonk8s.netease.com/tenant1-project1/spark:v3.0.0-20190813
spark.executor.instanses=5
spark.kubernetes.namespace=ns1
spark.kubernetes.container.image.pullSecrets=mysecret
spark.kubernetes.hadoop.configMapName=hz10-hadoop-dir
spark.kubernetes.kerberos.krb5.path=/etc/krb5.conf
spark.kerberos.principal=kyuubi/dev@HADOOP.HZ.NETEASE.COM
spark.kerberos.keytab=/Users/Kent/Downloads/kyuubi.keytab
```

### Does this PR introduce any user-facing change?

I guess this pr will now use config map of Hadoop from k8s cluster to the local client process if there are files to upload to Hadoop.

### How was this patch tested?

manually tested with spark + k8s cluster + standalone kerberized hdfs cluster
add an unit test

### CODE REVIEW

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/KubernetesUtils.scala [hunk 1]
```diff
@@ -267,30 +271,56 @@ private[spark] object KubernetesUtils extends Logging {
...
+  def getUploadPath(conf: SparkConf, client: KubernetesClient): (FileSystem, String) = {
+    conf.get(KUBERNETES_FILE_UPLOAD_PATH) match {
+      case Some(path) =>
+        val hadoopConf = new Configuration()
+        // When spark.kubernetes.file.upload.path is set, we need a cluster specific hadoop config,
+        // and if we use spark.kubernetes.hadoop.configMapName to configure not HADOOP_CONF_DIR, we
+        // should download the configmap to our client side.
+        // 1. add configurations from k8s configmap to hadoopConf
+        conf.get(KUBERNETES_HADOOP_CONF_CONFIG_MAP).foreach { cm =>
+          val hadoopConfFiles = client.configMaps().withName(cm).get().getData.asScala
```
Comment (skonto) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/KubernetesUtils.scala]: You can easily mount a configmap with the pod template feature no need to fetch the config map here.

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/KubernetesUtils.scala [hunk 2]
```diff
@@ -267,30 +271,56 @@ private[spark] object KubernetesUtils extends Logging {
...
           }
-        } else {
-          throw new SparkException("Please specify " +
-            "spark.kubernetes.file.upload.path property.")
         }
-      case _ => throw new SparkException("Spark configuration is missing...")
+        // 2. add configurations from arguments or spark properties file to hadoopConf
+        SparkHadoopUtil.appendS3AndSparkHadoopConfigurations(conf, hadoopConf)
+        // 3. set or rest user group information
+        UserGroupInformation.setConfiguration(hadoopConf)
```
Comment (skonto) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/KubernetesUtils.scala]: Spark submit should take care of the security stuff.

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/KubernetesUtils.scala (hunk 1 above)
Comment (yaooqinn) [reply resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/KubernetesUtils.scala]: 1. `spark.kubernetes.hadoop.configMapName` has no `driver` word in its name, so I guess it is ok to be used all over the spark application including the client process.
2. Using `spark.kubernetes.hadoop.configMapName` and `HADOOP_CONF_DIR` is an **either-or** thing, [check here](https://github.com/apache/spark/blob/5cf2602ccbcada92f11ac715872061f8307d9d70/resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/HadoopConfDriverFeatureStep.scala#L41).  It is more like that they both contain the same thing and are used by client to define the driver pod, so they should be equal. For now, the only difference between them is that the `HADOOP_CONF_DIR`  is in the classpath of client process and the other not.

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/KubernetesUtils.scala (hunk 2 above)
Comment (yaooqinn) [reply resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/KubernetesUtils.scala]: Yes, I agree. But for now, the spark submit is quite yarn-specific. For security stuff contains in configMap of k8s, it's not working.

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/KubernetesUtils.scala (hunk 1 above)
Comment (skonto) [reply resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/KubernetesUtils.scala]: I see that their either-or, but I think if you specify them both it should work.Have you tried that with HADOOP_CONF_DIR? Initially config map was not meant for uploading files from the client machine so it may need to be extended, but I do find fetching the configmap from the cluster redundant.

Comment (skonto) [reply resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/KubernetesUtils.scala]: @ifilonenko thoughts here? Do you think we can get the kerberos integration in here?

Comment (erikerlandson) [reply resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/KubernetesUtils.scala]: If they are both specified, the main corner case seems like the scenario where they aren't consistent, but this could be checked for.

Comment (yaooqinn) [reply resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/KubernetesUtils.scala]: I agree with setting them both, which is my original idea too.

Comment (ifilonenko) [reply resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/KubernetesUtils.scala]: @skonto ill make sure to get that PR resolved so that we can use it for testing, yeah

**Comment by yaooqinn** [comment]
gentle ping @skonto

**Comment by skonto** [comment]
@yaooqinn will review thanks

File: docs/security.md [hunk 3]
```diff
@@ -845,8 +845,13 @@ When talking to Hadoop-based services behind Kerberos, it was noted that Spark n
...
 
-In all cases you must define the environment variable: `HADOOP_CONF_DIR` or
-`spark.kubernetes.hadoop.configMapName.`
+In all cases you must define the environment variable: `HADOOP_CONF_DIR` or/and
+`spark.kubernetes.hadoop.configMapName`. When only `HADOOP_CONF_DIR` is configured, the hadoop configurations in there
+will be used by the client process, the Driver and its Executors. When only `spark.kubernetes.hadoop.configMapName`
+is configured, the hadoop configurations will only be used by the Driver and its Executors. If your client process has
+extra dependencies to upload to `spark.kubernetes.file.upload.path`, you may need to configure `HADOOP_CONF_DIR` too.
+When these two variables are both set, Spark will prefer `spark.kubernetes.hadoop.configMapName` to be mounted on the
+Driver/Executor pods.
```
Comment (skonto) [inline docs/security.md]: The concept looks good to me, @ifilonenko any corner cases?

File: resource-managers/kubernetes/core/src/test/scala/org/apache/spark/deploy/k8s/features/HadoopConfDriverFeatureStepSuite.scala [hunk 4]
```diff
@@ -62,6 +65,17 @@ class HadoopConfDriverFeatureStepSuite extends SparkFunSuite {
     assert(hadoopConfMap.getData().keySet().asScala === confFiles)
   }
 
+  test("using hadoop config map ahead of conf dir if both defined") {
+    val sparkConf = new SparkConfWithEnv(Map(ENV_HADOOP_CONF_DIR -> confDir.getAbsolutePath()))
+      .set(Config.KUBERNETES_HADOOP_CONF_CONFIG_MAP, "testConfigMap")
+
+    val conf = KubernetesTestConf.createDriverConf(sparkConf = sparkConf)
```
Comment (skonto) [inline resource-managers/kubernetes/core/src/test/scala/org/apache/spark/deploy/k8s/features/HadoopConfDriverFeatureStepSuite.scala]: nit: remove space

**Comment by skonto** [comment]
@erikerlandson LGTM, we need an integration test for this @ifilonenko gentle ping.

**Comment by yaooqinn** [comment]
anything blocks this？ @skonto

**Comment by yaooqinn** [comment]
retest this please

**Comment by yaooqinn** [comment]
retest this please

**Comment by yaooqinn** [comment]
cc @cloud-fan @vanzin, could any active committers help review this? thanks in advance.

**Comment by yaooqinn** [comment]
retest this please

**Comment by erikerlandson** [comment]
Does this have an integration test now?
cc @ifilonenko @skonto

File: docs/security.md (hunk 3 above)
Comment (ScrapCodes) [reply docs/security.md]: I think it is still okay to warn the user about, the configuration is being picked from the configMap, inspite of the fact, HADOOP_CONF_DIR is defined.

**Comment by yaooqinn** [comment]
retest this please
